cluster-autoscaler/core/scaledown/actuation/priority_evictor.go

```diff
@@ -0,0 +1,113 @@
+type PriorityEvictor struct {
+	EvictionRetryTime                time.Duration
+	DsEvictionRetryTime              time.Duration
+	DsEvictionEmptyNodeTimeout       time.Duration
+	PodEvictionHeadroom              time.Duration
+	evictionRegister                 evictionRegister
+	deleteOptions                    simulator.NodeDeleteOptions
+	shutdownGracePeriodByPodPriority []kubelet_config.ShutdownGracePeriodByPodPriority
+}
+
+// NewPriorityEvictor returns an instance of PriorityEvictor using the default parameters.
+func NewPriorityEvictor(deleteOptions simulator.NodeDeleteOptions, evictionRegister evictionRegister, shutdownGracePeriodByPodPriority []kubelet_config.ShutdownGracePeriodByPodPriority) PriorityEvictor {
+	sort.Slice(shutdownGracePeriodByPodPriority, func(i, j int) bool {
+		return shutdownGracePeriodByPodPriority[i].Priority < shutdownGracePeriodByPodPriority[j].Priority
+	})
+
+	return PriorityEvictor{
+		EvictionRetryTime:                DefaultEvictionRetryTime,
+		DsEvictionRetryTime:              DefaultDsEvictionRetryTime,
+		DsEvictionEmptyNodeTimeout:       DefaultDsEvictionEmptyNodeTimeout,
+		PodEvictionHeadroom:              DefaultPodEvictionHeadroom,
+		evictionRegister:                 evictionRegister,
+		deleteOptions:                    deleteOptions,
+		shutdownGracePeriodByPodPriority: shutdownGracePeriodByPodPriority,
+	}
+}
+
+// DrainNode groups pods in the node in to priority groups and, evicts pods in the ascending order of priorities.
+// All non-mirror, non-DS pods on the node are evicted. Mirror pods are not evicted.
+// DaemonSet pods are evicted if DaemonSetEvictionForOccupiedNodes is enabled, or if they have the EnableDsEvictionKey annotation.
+func (e PriorityEvictor) DrainNode(ctx *context.AutoscalingContext, nodeInfo *framework.NodeInfo) (map[string]status.PodEvictionResult, error) {
+	pods, dsPods := podsToEvict(ctx, nodeInfo)
+	return e.DrainNodeWithPodsBasedOnPodPriority(ctx, nodeInfo.Node(), append(pods, dsPods...))
+}
+
+// EvictDaemonSetPods creates eviction objects for all DaemonSet pods on the node.
+func (e PriorityEvictor) EvictDaemonSetPods(ctx *context.AutoscalingContext, nodeInfo *framework.NodeInfo) error {
+	return evictDaemonSetPods(ctx, nodeInfo, e.deleteOptions, e.DsEvictionEmptyNodeTimeout, e.DsEvictionRetryTime, e.evictionRegister)
+}
+
+// DrainNodeWithPodsBasedOnPodPriority performs drain logic on the node based on pod priorities. Marks the node as unschedulable and
+// later removes all pods, giving them up to MaxGracefulTerminationTime to finish. The list of pods to evict has to be provided.
+func (e PriorityEvictor) DrainNodeWithPodsBasedOnPodPriority(ctx *context.AutoscalingContext, node *apiv1.Node, pods []*apiv1.Pod) (map[string]status.PodEvictionResult, error) {
+	return drainNodeWithPodsBasedOnPodPriority(ctx, node, pods, e.shutdownGracePeriodByPodPriority, false, e.EvictionRetryTime, e.evictionRegister, e.PodEvictionHeadroom)
```

Comment:
Am I getting it right that the only difference between the two Evictors is the `ignoreDaemonSetPodEvictionResult` parameter to `drainNodeWithPodsBasedOnPodPriority`? If that's true, maybe we should merge the two implementations after all, and just parameterize it with whether it should evict DS pods fully, or best-effort? If a user configures the old flag, in initialization we'd turn it into the single priority group + pass "best-effort DS eviction" to the evictor. If a user configures the new flag, we'd just parse the groups + pass "full DS eviction". 

That would remove a lot of redundancy (the struct params, constructor etc), and we could keep the logic from drain.go as methods, so we'd pass fewer things by arguments. WDYT?

Reply:
Done